compiler/rustc_mir/src/transform/instcombine.rs

```diff
@@ -87,6 +93,58 @@ impl OptimizationFinder<'b, 'tcx> {
         OptimizationFinder { body, tcx, optimizations: OptimizationList::default() }
     }
 
+    fn find_deref_of_address(&mut self, rvalue: &Rvalue<'tcx>, location: Location) -> Option<()> {
+        // look for the sequence
+        // _2 = &_1;
+        // ...
+        // _5 = (*_2)
+        //
+        // which we can replace the last statement with `_5 = _1` to avoid the load of _2
+        if let Rvalue::Use(op) = rvalue {
+            let local_being_derefed = match op.place()?.as_ref() {
+                PlaceRef { local, projection: [ProjectionElem::Deref] } => Some(local),
+                _ => None,
+            }?;
+
+            let stmt_index = location.statement_index;
+            // look behind for statement that assigns the local from a address of operator
+            // 6 is chosen as a heuristic determined by seeing the number of times the optimization kicked in compiling rust std
+            let lower_index = stmt_index.saturating_sub(6);
+            let statements_to_look_in = self.body.basic_blocks()[location.block].statements
+                [lower_index..stmt_index]
+                .iter()
+                .rev();
+            for stmt in statements_to_look_in {
+                match &stmt.kind {
+                    rustc_middle::mir::StatementKind::Assign(box (l, r))
+                        if l.local == local_being_derefed =>
+                    {
+                        match r {
+                            // looking for _local_being_deref = &_1;
+                            Rvalue::Ref(
+                                _,
+                                BorrowKind::Shared | BorrowKind::Shallow,
+                                place_taken_address_of,
+                            ) => {
+                                self.optimizations
+                                    .unneeded_deref
+                                    .insert(location, *place_taken_address_of);
+                                return Some(());
+                            }
+                            _ => {
+                                // we found an assignment of the place that is not a ref, e.g _local_being_derefed = _2;
+                                // We therefore have to bail the optimization since the optimization is no longer valid
+                                return None;
+                            }
+                        }
+                    }
+                    _ => {}
```

Comment:
if `local_being_derefed` is used on the rhs of an assignment to another local, your optimization may misfire I think.

```rust
    let x = 42;
    let a = 99;
    let mut y = &x;
    let z = &mut y;
    *z = &a;
    println!("{}", *y);
```

will not print `99` after your optimization but `42`

Reply:
I added a test for this. It does not seem like it will misfire. I also added more clarifying comments on the matches. I have persuaded myself that since we are only applying this optimization on immutable references, we can't have a mutable reference at the same time, so nothing (besides asm) can break this optimization. My reasoning may be wrong though..

Reply:
You are correct. Setting the lookback higher does indeed cause a miscompilation. I'll look into fixing it.

Reply:
The general way to look for mutation is to define a visitor and visit the construct you want to analyze (`Statement` in your case) and implement https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/visit/trait.Visitor.html#method.visit_local to and check the `PlaceContext` if the local matches the local you want to check.

Reply:
Ah, I saw your comment a tad too late. I already pushed a fix, but i'll see if it is cleaner/less adhoc to use the visitor. Thanks for the hint.

Reply:
The latest commits now check that local_being_derefed is not mutated using a visitor. I have verified that setting the lookback to 10000 causes a misoptimization without the fix, and fixes it with the fix. I have set it back to to avoid bad runtime complexity. I'm not sure how best to represent that the fix is actually working. For now I separated it into two commits.

Reply:
Thanks! Splitting it into individual commits made this great to review.

I also don't know how to test it. A bogus idea (Don't do this :laughing:): instead of `6` use `6 + mir_opt_level`, then we can do `-Zmir-opt-level=99999` on that test.

An alternative to having this magic number would be to implement the optimization as a feed forward (am I using words correctly here?) optimization. What I mean is that we don't go through every deref and look back, but we walk forward through each block and keep a set of "deref-optimizable" locals that we update as we go.

Reply:
That said, I'm fine with merging this optimization as long as there's a tracking issue for exploring such a change to the optimization. I'm not sure whether that change is feasible in practice, but we should explore it

Reply:
Cool. Yeah it should be possible to do this optimization  without the lookback, such that a single statement is only visited once. I'll open a new pr for that.

Can we do a perf run on the current pr? I'm curious if this has any impact